text/2018-09-17-schedule-limit.md

```diff
@@ -0,0 +1,176 @@
+a short period of time and affecting the performance of the TiKV cluster, and
+to avoid the problem of over-balance, currently PD uses a limitation based
+scheme.
+
+Specifically, PD uses 4 different limit configurations to control the number of
+concurrently running operators:
+
+* _leader-schedule-limit_
+
+    Controls the number of operators scheduling region leaders at the same
+    time.
+
+* _region-schedule-limit_
+
+    Controls the number of operators scheduling region peers at the same time.
+
+* _replica-schedule-limit_
+
+    Controls the number of operators scheduling region replicas at the same
+    time.
+
+* _merge-schedule-limit_
+
+    Controls the number of region merge operators at the same time.
+
+![Old schedule limit scheme](../media/schedule-limit-old.png)
+
+The mechanism of scheduling limiter is shown in the figure. All running
+operators are maintained as a map in the coordinator, and the Limiter is
+updated synchronously when running operators change. Different schedulers will
+check whether running operators have exceeded the limit before generating
+operators. The operators generated by schedulers will be added to the
+_runningOperators_ map via addOperator method.
+
+This scheme is not good in these aspects:
+
+* Various limit configurations are confusing, tricky to use correctly
+* Fixed limit configuration does not apply to clusters of different sizes
+* The preemption strategy between different schedulers is not fair enough
+* Scheduling is often concentrated on a small number of `tikv-server`s, which
+  leads to slow scheduling sometimes and may affect the latency of TiKV (we do
+  consider snapshot count, but this statistic is reported after TiKV receives
+  the scheduling request)
+* Rely on schedulers correctly checking limit to ensure the limit configuration
+  has not been violated -- hotbed of bugs. See
+  [#1193](https://github.com/pingcap/pd/pull/1193)
+  [#1155](https://github.com/pingcap/pd/pull/1155).
+
+## Detailed design
+
+### Introducing waitingOperators queue
+
+![New schedule limit scheme](../media/schedule-limit-new.png)
+
+As shown in the figure, the biggest change is the introduction of the
+_waitingOperators_ in the coordinator. The schedule operators generated by
+schedulers are not directly put into the _runningOperators_, but are queued
+first in _waitingOperators_. The coordinator is responsible for continuously
+promoting operators in _waitingOperators_ queue to _runningOperators_ according
+to a certain rule and configuration.
+
+### Operator effect evaluation
+
+Schedule operators consist of operator steps. The effect of an operator on the
+cluster can be calculated as the sum of the effects of all steps. There are
+several types of operator steps:
+
+* _TransferLeader_
+* _AddPeer_
+* _AddLearner_
+* _PromoteLearner_
+* _RemovePeer_
+* _MergeRegion_
+* _SplitRegion_
+
+Obviously, these steps only affect a subset of all `tikv-server`s -- not the
```

Comment:
```suggestion
These steps only affect a subset of all `tikv-server`s -- not the
```

"Obviously" adds nothing to this sentence other than possibly make some readers feel stupid.